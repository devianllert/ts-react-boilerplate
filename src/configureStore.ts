import {
  createStore,
  compose,
  applyMiddleware,
  Store,
} from 'redux';
import { routerMiddleware } from 'connected-react-router';

import createReducer from './reducers';

const configureStore = (initialState = {}, history: import('history').History): Store => {
  let composeEnhancers = compose;

  // If Redux Dev Tools and Saga Dev Tools Extensions are installed, enable them
  if (process.env.NODE_ENV !== 'production' && typeof window === 'object') {
    /* eslint-disable */
    // @ts-ignore
    if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__)
      // @ts-ignore
      composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({});

    // NOTE: Uncomment the code below to restore support for Redux Saga
    // Dev Tools once it supports redux-saga version 1.x.x
    // if (window.__SAGA_MONITOR_EXTENSION__)
    //   reduxSagaMonitorOptions = {
    //     sagaMonitor: window.__SAGA_MONITOR_EXTENSION__,
    //   };
    /* eslint-enable */
  }

  const middlwares = [
    routerMiddleware(history),
  ];

  const enhancers = [applyMiddleware(...middlwares)];

  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(...enhancers),
  );

  return store;
};

export default configureStore;
