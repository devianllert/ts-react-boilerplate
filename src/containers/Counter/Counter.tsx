import React, { ReactElement } from 'react';
import { connect } from 'react-redux';

import { counterIncrement, counterDecrement } from './actions';

import Button from '../../components/Button';

interface Props {
  count: number;
  increment: () => import('./types').CounterIncrement;
  decrement: () => import('./types').CounterDecrement;
}

const Counter = ({ count, increment, decrement }: Props): ReactElement => (
  <div>
    <Button onClick={increment}>+</Button>
    <span>{count}</span>
    <Button onClick={decrement}>-</Button>
  </div>
);

const mapStateToProps = (state: any): any => ({
  count: state.counter.count,
});

const mapDispatchToProps = {
  increment: counterIncrement,
  decrement: counterDecrement,
};

export default connect<Props, typeof mapDispatchToProps, {}>(mapStateToProps, mapDispatchToProps)(Counter);
