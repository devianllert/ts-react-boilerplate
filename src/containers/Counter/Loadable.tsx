import loadable, { ImportComponent } from '../../utils/loadable';

const LoadableCounter = loadable((): ImportComponent => import(/* webpackChunkName: "counter" */ './Counter'));

export default LoadableCounter;
