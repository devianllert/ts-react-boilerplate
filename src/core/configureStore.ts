import {
  createStore,
  compose,
  applyMiddleware,
  Store,
  Reducer,
} from 'redux';
import createSagaMiddleware, { Saga, SagaMiddleware, Task } from 'redux-saga';

import createReducer from './reducers';

export interface EnhancedStore extends Store {
  runSaga: SagaMiddleware['run'];
  injectedReducers: {
    [key: string]: Reducer;
  };
  injectedSagas: {
    [key: string]: {
      saga?: Saga;
      task: Task;
      mode?: string;
    } | string;
  };
}

const configureStore = (initialState = {}): EnhancedStore => {
  let composeEnhancers = compose;
  const reduxSagaMonitorOptions = {};

  // If Redux Dev Tools and Saga Dev Tools Extensions are installed, enable them
  if (process.env.NODE_ENV !== 'production' && typeof window === 'object') {
    /* eslint-disable */
    // @ts-ignore
    if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) {
      // @ts-ignore
      composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({});
    }
    /* eslint-enable */
  }

  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);

  const middlwares = [
    sagaMiddleware,
  ];

  const enhancers = [applyMiddleware(...middlwares)];

  const store: EnhancedStore = createStore(
    createReducer(),
    initialState,
    composeEnhancers(...enhancers),
  );

  // Extensions
  store.runSaga = sagaMiddleware.run;
  store.injectedReducers = {}; // Reducer registry
  store.injectedSagas = {}; // Saga registry

  return store;
};

export default configureStore;
