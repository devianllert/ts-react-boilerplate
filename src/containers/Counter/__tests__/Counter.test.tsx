import React from 'react';
import { Provider } from 'react-redux';
import { cleanup, render, fireEvent } from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect';

import configureStore from '../../../configureStore';

import Counter from '../Counter';

describe('<Counter />', (): void => {
  afterEach(cleanup);

  const store = configureStore();

  it('should render and match the snapshot', (): void => {
    const { container } = render(
      <Provider store={store}>
        <Counter />
      </Provider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should increment count', (): void => {
    const { getByText } = render(
      <Provider store={store}>
        <Counter />
      </Provider>,
    );

    expect(getByText(/0/)).toBeDefined();

    fireEvent.click(getByText('+'));

    expect(getByText(/1/)).toBeDefined();
  });
});
