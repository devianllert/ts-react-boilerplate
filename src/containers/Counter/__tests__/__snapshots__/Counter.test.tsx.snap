// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> should render and match the snapshot 1`] = `
<div>
  <button
    class="button"
    type="button"
  >
    +
  </button>
  <span>
    0
  </span>
  <button
    class="button"
    type="button"
  >
    -
  </button>
</div>
`;
